### PR TITLE
10020: Remove ringbuffer from the head tracker

### PR DESCRIPTION
**Pull request by archseer**
We now have head sampling, and all subscribers of OnNewLongestChain are async, so the head buffer is no longer needed

### CODE REVIEW

File: core/services/head_tracker.go [hunk 1]
```diff
@@ -505,7 +413,7 @@ func (ht *HeadTracker) receiveHeaders(ctx context.Context) error {
 				deadlineCtx, cancel := context.WithTimeout(ctx, timeBudget)
 				defer cancel()
 
-				err := ht.handleNewHead(ctx, blockHeader)
+				err := ht.handleNewHead(ctx, *blockHeader)
```
Comment (PiotrTrzpil) [inline core/services/head_tracker.go:416]: previously the ring buffer had:
```
			if h == nil {
				r.logger().Error("HeadTracker: got nil block header")
				continue
			}
```
so might be good to keep it

File: core/services/head_tracker.go [hunk 2]
```diff
@@ -25,14 +25,6 @@ var (
 		Name: "head_tracker_current_head",
 		Help: "The highest seen head number",
 	})
-	promNumHeadsReceived = promauto.NewCounter(prometheus.CounterOpts{
```
Comment (PiotrTrzpil) [inline core/services/head_tracker.go:28]: worth keeping this counter still

**Review by PiotrTrzpil — COMMENTED** [review]
looks good overall 👍🏼

File: core/services/head_tracker.go [hunk 3]
```diff
@@ -496,7 +408,7 @@ func (ht *HeadTracker) receiveHeaders(ctx context.Context) error {
 		select {
 		case <-ht.done:
 			return nil
-		case blockHeader, open := <-ht.outHeaders:
+		case blockHeader, open := <-ht.headers:
 			if !open {
 				return errors.New("HeadTracker: outHeaders prematurely closed")
 			}
```
Comment (PiotrTrzpil) [inline core/services/head_tracker.go:414]: hm i wonder if this closing behaviour should be different now. In the ringbuffer we had:
```
if !ok {
				// If the channel is closed, do not handle any more heads.
				close(r.out)
				return
			}
```
Though it was also closing `r.out` so we would still result in `return errors.New("HeadTracker: outHeaders prematurely closed")` unless `case <-ht.done:` was received first.
Maybe it's fine, but would be good to test this disconnection

File: core/services/head_tracker.go [hunk 4]
```diff
@@ -340,6 +252,9 @@ func (ht *HeadTracker) listenForNewHeads() {
 			break
 		} else if err != nil {
 			ht.logger().Errorw("Error in new head subscription, unsubscribed", "err", err)
+			// subscribe() will call unsubscribeFromHead() which will panic
+			// closing the channel if already closed.
+			ht.headers = make(chan *models.Head)
```
Comment (PiotrTrzpil) [inline core/services/head_tracker.go:257]: hm so it was being double-closed? Or perhaps was panicking on nil? Not sure if the comment is explanatory

In `unsubscribeFromHead` we also have 
```

	if !hl.connected {
		return nil
	}
```

so it should not attempt to close the channel if we are not connected. Unless the channel can be closed by the geth client? That would be similar to what you have in the test

**Review by PiotrTrzpil — APPROVED** [review]
nice one 👍🏼 @samsondav @spooktheducks do you also have a moment to look?

**Review by samsondav — APPROVED** [review]
Great work

**Comment by archseer** [comment]
(Force pushed since github's merge status got stuck)

**Review by PiotrTrzpil — APPROVED** [review]

**Review by spooktheducks — APPROVED** [review]

**Pull request merged**